spec/unit/driver/auto_scaling_group_spec.rb

```diff
@@ -131,4 +135,47 @@
       expect(@driver.instance_statuses).to eq( { 'i-abcd1234' => returned_status } )
     end
   end
+
+  describe '#wait_for_desired_capacity' do
+    it 'completes if desired capacity reached' do
+      expect(@driver).to receive(:desired_capacity_reached?).and_return(true)
+
+      @driver.wait_for_desired_capacity
+    end
+
+    it 'times out if desired capacity is not reached' do
+      expect(@driver).to receive(:desired_capacity_reached?).and_return(false)
+
+      expect { @driver.wait_for_desired_capacity }.to raise_error(Timeout::Error)
+    end
+  end
+
+  describe '#desired_capacity_reached?' do
+    it 'returns true if healthy instance count matches desired capacity' do
+      expected_number = 5
+
+      expect(group).to receive(:desired_capacity).and_return(expected_number)
+      expect(@driver).to receive(:healthy_instance_count).and_return(expected_number)
+
+      expect(@driver.desired_capacity_reached?).to be_true
+    end
+
+    it 'returns false if healthy instance count is less than desired capacity' do
+      expected_number = 5
+
+      expect(group).to receive(:desired_capacity).and_return(expected_number)
+      expect(@driver).to receive(:healthy_instance_count).and_return(expected_number - 1)
+
+      expect(@driver.desired_capacity_reached?).to be_false
+    end
+
+    it 'returns true if healthy instance count is more than desired capacity' do
```

Comment:
I'll preface that this is a contrived scenario.

There's a timing issue here in the real world.  We sleep for 15 seconds between each check.  Say on the first check, the desired instances is 5.  We have less than 5, so we sleep, waiting for auto scaling to kick in.  Auto scaling kicks in, and gets us to 5 instances, but as that happens, load drops, and our alarm sets our desired instances to 4, but auto-scaling hasn't kicked in yet to scale down that 5th instance.  All this happens within our 15 second sleep window.  We wake up, see that we want 4 instances but have 5.  Should we sleep again, or say that this is acceptable, and pass our warm_up?

Since the intent of warm_up is to ensure equivalent capacity, I think we should say that this is acceptable.